pkg/localstore/gc_test.go

```diff
@@ -534,3 +535,194 @@ func TestSetTestHookCollectGarbage(t *testing.T) {
+		if err != nil {
+			fmt.Println("Pinned chunk missing ", addr)
+			t.Fatal(err)
+		}
+		if !gotChunk.Address().Equal(swarm.NewAddress(addr.Bytes())) {
+			t.Fatal("Pinned chunk is not equal to got chunk")
+		}
+	}
+
+}
+
+func generateAndPinAChunk(t *testing.T, db *DB) swarm.Chunk {
+	// Create a chunk and pin it
+	pinnedChunk := generateTestRandomChunk()
+
+	_, err := db.Put(context.Background(), storage.ModePutUpload, pinnedChunk)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = db.Set(context.Background(), storage.ModeSetPin, pinnedChunk.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = db.Set(context.Background(), storage.ModeSetSyncPull, pinnedChunk.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+	return pinnedChunk
+}
+
+func TestPinSyncAndAccessPutSetChunkMultipleTimes(t *testing.T) {
+	var closed chan struct{}
+	testHookCollectGarbageChan := make(chan uint64)
+	t.Cleanup(setTestHookCollectGarbage(func(collectedCount uint64) {
+		select {
+		case testHookCollectGarbageChan <- collectedCount:
+		case <-closed:
+		}
+	}))
+	db := newTestDB(t, &Options{
+		Capacity: 10,
+	})
+	closed = db.close
+
+	rch := generateTestRandomChunk()
+	for i := 0; i < 20; i++ {
+		ch := generateTestRandomChunk()
+
+		// pin a chunk in between
+		if i == 5 {
+			_, err := db.Put(context.Background(), storage.ModePutUpload, rch)
+			if err != nil {
+				t.Fatal(err)
+			}
+			err = db.Set(context.Background(), storage.ModeSetPin, rch.Address())
+			if err != nil {
+				t.Fatal(err)
+			}
+			continue
+		}
+
+		_, err := db.Put(context.Background(), storage.ModePutUpload, ch)
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = db.Set(context.Background(), storage.ModeSetSyncPull, ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = db.Set(context.Background(), storage.ModeSetAccess, ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = db.Get(context.Background(), storage.ModeGetRequest, ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	_, err := db.Get(context.Background(), storage.ModeGetRequest, rch.Address())
+	if err != nil {
+		fmt.Println("Pinned chunk missing ", rch.Address())
+		t.Fatal(err)
+	}
+
+	addRandomChunks(t, 20, db)
+
+	_, err = db.Get(context.Background(), storage.ModeGetRequest, rch.Address())
+	if err != nil {
+		fmt.Println("Pinned chunk missing ", rch.Address())
+		t.Fatal(err)
+	}
+
+	addRandomChunks(t, 20, db)
+
+	_, err = db.Get(context.Background(), storage.ModeGetRequest, rch.Address())
+	if err != nil {
+		fmt.Println("Pinned chunk missing ", rch.Address())
+		t.Fatal(err)
+	}
+
+	// check if the pinned chunk is present after GC
+	gotChunk, err := db.Get(context.Background(), storage.ModeGetRequest, rch.Address())
+	if err != nil {
+		t.Fatal("Pinned chunk missing ", err)
+	}
+	if !gotChunk.Address().Equal(rch.Address()) {
+		t.Fatal("Pinned chunk is not equal to got chunk")
+	}
+
+}
+
+func addRandomChunks(t *testing.T, count int, db *DB) {
```

Comment:
why don't you use this in all of the tests above?

Reply:
Because other tests dont need this.